File: 0.36/mame_src/src/sndhrdw/invinco.c
/*
 *	Invinco sound routines
 */

#include "driver.h"


/* output port 0x02 definitions - sound effect drive outputs */
#define OUT_PORT_2_SAUCER		0x04
#define OUT_PORT_2_MOVE1		0x08
#define OUT_PORT_2_MOVE2		0x10
#define OUT_PORT_2_FIRE			0x20
#define OUT_PORT_2_INVHIT		0x40
#define OUT_PORT_2_SHIPHIT		0x80


#define PLAY(id,loop)           sample_start( id, id, loop )
#define STOP(id)                sample_stop( id )


/* sample file names */
const char *invinco_sample_names[] =
{
	"*invinco",
	"saucer.wav",
	"move1.wav",
	"move2.wav",
	"fire.wav",
	"invhit.wav",
	"shiphit.wav",
	"move3.wav",	/* currently not used */
	"move4.wav",	/* currently not used */
	0
};

/* sample sound IDs - must match sample file name table above */
enum
{
	SND_SAUCER = 0,
	SND_MOVE1,
	SND_MOVE2,
	SND_FIRE,
	SND_INVHIT,
	SND_SHIPHIT,
	SND_MOVE3,
	SND_MOVE4
};


void invinco_sh_port2_w( int offset, int data )
{
	static int port2State = 0;
	int bitsChanged;
	int bitsGoneHigh;
	int bitsGoneLow;


	bitsChanged  = port2State ^ data;
	bitsGoneHigh = bitsChanged & data;
	bitsGoneLow  = bitsChanged & ~data;

	port2State = data;

	if ( bitsGoneLow & OUT_PORT_2_SAUCER )
	{
		PLAY( SND_SAUCER, 0 );
	}

	if ( bitsGoneLow & OUT_PORT_2_MOVE1 )
	{
		PLAY( SND_MOVE1, 0 );
	}

	if ( bitsGoneLow & OUT_PORT_2_MOVE2 )
	{
		PLAY( SND_MOVE2, 0 );
	}

	if ( bitsGoneLow & OUT_PORT_2_FIRE )
	{
		PLAY( SND_FIRE, 0 );
	}

	if ( bitsGoneLow & OUT_PORT_2_INVHIT )
	{
		PLAY( SND_INVHIT, 0 );
	}

	if ( bitsGoneLow & OUT_PORT_2_SHIPHIT )
	{
		PLAY( SND_SHIPHIT, 0 );
	}

#if 0
	if (errorlog) fprintf(errorlog,"Went LO: %02X  %04X\n", bitsGoneLow, cpu_get_pc());
#endif
}

File: 0.36/emulator/src/main/java/arcadeflex/v036/sndhrdw/invinco.java
/*
 * ported to v0.36
 */
/**
 * Changelog
 * =========
 * 14/01/2023 - shadow - This file should be complete for 0.36 version
 */
package arcadeflex.v036.sndhrdw;

//generic imports
import static arcadeflex.v036.generic.funcPtr.*;
//sound imports
import static arcadeflex.v036.sound.samples.*;

public class invinco {

    /* output port 0x02 definitions - sound effect drive outputs */
    public static final int OUT_PORT_2_SAUCER = 0x04;
    public static final int OUT_PORT_2_MOVE1 = 0x08;
    public static final int OUT_PORT_2_MOVE2 = 0x10;
    public static final int OUT_PORT_2_FIRE = 0x20;
    public static final int OUT_PORT_2_INVHIT = 0x40;
    public static final int OUT_PORT_2_SHIPHIT = 0x80;

    public static void PLAY(int id, int loop) {
        sample_start(id, id, loop);
    }

    public static void STOP(int id) {
        sample_stop(id);
    }

    /* sample file names */
    public static String invinco_sample_names[]
            = {
                "*invinco",
                "saucer.wav",
                "move1.wav",
                "move2.wav",
                "fire.wav",
                "invhit.wav",
                "shiphit.wav",
                "move3.wav", /* currently not used */
                "move4.wav", /* currently not used */
                null
            };

    /* sample sound IDs - must match sample file name table above */
    public static final int SND_SAUCER = 0;
    public static final int SND_MOVE1 = 1;
    public static final int SND_MOVE2 = 2;
    public static final int SND_FIRE = 3;
    public static final int SND_INVHIT = 4;
    public static final int SND_SHIPHIT = 5;
    public static final int SND_MOVE3 = 6;
    public static final int SND_MOVE4 = 7;

    static int port2State = 0;
    public static WriteHandlerPtr invinco_sh_port2_w = new WriteHandlerPtr() {
        public void handler(int offset, int data) {
            int bitsChanged;
            int bitsGoneHigh;
            int bitsGoneLow;

            bitsChanged = port2State ^ data;
            bitsGoneHigh = bitsChanged & data;
            bitsGoneLow = bitsChanged & ~data;

            port2State = data;

            if ((bitsGoneLow & OUT_PORT_2_SAUCER) != 0) {
                PLAY(SND_SAUCER, 0);
            }

            if ((bitsGoneLow & OUT_PORT_2_MOVE1) != 0) {
                PLAY(SND_MOVE1, 0);
            }

            if ((bitsGoneLow & OUT_PORT_2_MOVE2) != 0) {
                PLAY(SND_MOVE2, 0);
            }

            if ((bitsGoneLow & OUT_PORT_2_FIRE) != 0) {
                PLAY(SND_FIRE, 0);
            }

            if ((bitsGoneLow & OUT_PORT_2_INVHIT) != 0) {
                PLAY(SND_INVHIT, 0);
            }

            if ((bitsGoneLow & OUT_PORT_2_SHIPHIT) != 0) {
                PLAY(SND_SHIPHIT, 0);
            }
        }
    };
}
